File: matlab/read_pillar_bin.m
%read_pillar_bin Read PILLAR binary data.
%   [META, DATA] = read_pillar_bin(FILENAME) reads binary data from the
%   specified file by the string FILENAME. The return value META stores
%   file meta information, parameters used to generate this file. and DATA
%   stores the raw coordinates, drift-corrected coordinates and
%   corresponding deflections if existed.
function [file_header, data] = read_pillar_bin(fname)
    file_header = struct;
    data = struct;
    
    %% open the file
    fileID = fopen(fname);
    readmode = 'b';
    
    %% read header
    fileversion1 = 589450;  %file version control
    fileversion2 = 589451;  %since v1.1.3 if grid restoration applied
    file_header.version = fread(fileID,1,'int',readmode);
    file_header.npillars = 0;
    file_header.nframes = 0;
    file_header.suc_load = false;
    if(file_header.version==fileversion1 || file_header.version==fileversion2)
        file_header.npillars    = fread(fileID,1,'int',readmode);
        file_header.nframes     = fread(fileID,1,'int',readmode);
        file_header.pixel_size  = fread(fileID,1,'double',readmode);
        file_header.diameter    = fread(fileID,1,'double',readmode);
        file_header.spacing     = fread(fileID,1,'double',readmode);
        file_header.oblique     = fread(fileID,1,'double',readmode);
        file_header.grid_angle  = fread(fileID,1,'double',readmode);
        file_header.sigma_PSF   = fread(fileID,1,'double',readmode);
        file_header.catch_radius= fread(fileID,1,'double',readmode);
        file_header.kernel_w    = fread(fileID,1,'int',readmode);
        file_header.box_limit   = fread(fileID,1,'int',readmode);
        file_header.dark_pillar     = fread(fileID,1,'int8',readmode);
        file_header.use_minimum_std = fread(fileID,1,'int8',readmode);
        file_header.use_metric_CG   = fread(fileID,1,'int8',readmode);
        file_header.use_enhancer    = fread(fileID,1,'int8',readmode);  
        if(file_header.version==fileversion2)
            file_header.use_fft     = fread(fileID,1,'int8',readmode);  
        else
            file_header.use_fft = 0;
        end
        file_header.enhance_loaded_suc = true;
        if(file_header.use_enhancer>0)
            file_header.enhancer.normalized_cc = fread(fileID,1,'int8',readmode);
            file_header.enhancer.use_gauss_psf = fread(fileID,1,'int8',readmode);
            file_header.enhancer.gauss_psf_dark= fread(fileID,1,'int8',readmode);
            file_header.enhancer.gauss_psf_radius= fread(fileID,1,'int',readmode);
            file_header.enhancer.gauss_psf_sigma = fread(fileID,1,'double',readmode);
            file_header.enhancer.psf_w = fread(fileID,1,'int',readmode);
            file_header.enhancer.psf_h = fread(fileID,1,'int',readmode);        
            if(file_header.enhancer.psf_w>0 && file_header.enhancer.psf_h>0)
                file_header.enhancer.PSF = fread(fileID,[file_header.enhancer.psf_w, file_header.enhancer.psf_h],'double',readmode);            
            else
                file_header.enhance_loaded_suc = false;
            end
        end
        if(file_header.use_fft>0)
            file_header.FFT.off_center_radius               = fread(fileID,1,'int',readmode);
            file_header.FFT.center_radius                   = fread(fileID,1,'int',readmode);
            file_header.FFT.start_off_center_radius         = fread(fileID,1,'int',readmode);
            file_header.FFT.end_off_center_center_radius    = fread(fileID,1,'int',readmode);
            file_header.FFT.num_points                      = fread(fileID,1,'int',readmode);
            file_header.FFT.points                          = fread(fileID,[2,file_header.FFT.num_points],'int',readmode);
        end
        if(file_header.enhance_loaded_suc)
            startXY= fread(fileID,[2,file_header.npillars],'int',readmode); 
            file_header.startX = startXY(1,:);
            file_header.startY = startXY(2,:);
            file_header.suc_load = true;
        end
    elseif(mod(file_header.version,2)==0)
        file_header.npillars = version/2;
        file_header.nframes = fread(fileID,1,'int',readmode);
        file_header.suc_load = true;
    end

    %% check whether the file header is loaded successfully.
    if(file_header.suc_load && file_header.npillars>0 && file_header.nframes>0)
        %% read out the tracks
        ncols = file_header.npillars*2;
        nframes = file_header.nframes;
        A = fread(fileID,[ncols, nframes],'double',readmode);
        B = fread(fileID,[ncols, nframes],'double',readmode);
        if(file_header.version==fileversion2)
            %% deflections 
            data.DX=B(1:2:ncols,:)';
            data.DY=B(2:2:ncols,:)';
            %% drift corrected data
            B = fread(fileID,[ncols, nframes],'double',readmode);
        end

        %% raw data
        data.tracksX=A(1:2:ncols,:)';
        data.tracksY=A(2:2:ncols,:)';

        %% drift correct data
        if(numel(B)>0)
            data.drift_correctX=B(1:2:ncols,:)';
            data.drift_correctY=B(2:2:ncols,:)';
            % drift information
            driftXY = fread(fileID,[2, nframes],'double',readmode);
            data.driftX = driftXY(1,:);
            data.driftY = driftXY(2,:);
            % reference pillars flags 
            reference_flags = fread(fileID,file_header.npillars,'int8',readmode);
            data.reference = (reference_flags>0);
        end
    else
        file_header.suc_load = false;
        disp('loading file failed.');
    end

    %% colse the file handle
    fclose(fileID);
end

File: python/read_pillar_bin.py
import numpy as np

# load the pillar binary file, and wrap up into a class.
class PillarTrackerBinReader:
    fileversion1 = 589450  # file version control
    fileversion2 = 589451  # since v1.1.3 if grid restoration applied
    dt_int = np.dtype(np.int32).newbyteorder('>')
    dt_double = np.dtype(np.float64).newbyteorder('>')
    dt_int8 = np.dtype(np.int8).newbyteorder('>')

    def loadfile(self, fpath):
        fileversion1 = self.fileversion1
        fileversion2 = self.fileversion2
        dt_int = self.dt_int
        dt_double = self.dt_double
        dt_int8 = self.dt_int8

        class Meta:
            pass

        class ParaKernel:
            pass

        class ParaFFT:
            pass

        class Data:
            pass

        header = Meta()
        kernel = ParaKernel()
        fft = ParaFFT()
        data = Data()

        suc_load = False
        npillars = 0
        nframes = 0
        with open(fpath, "rb") as f:
            version = np.fromfile(f, dtype=dt_int, count=1)
            if version == fileversion1 or version == fileversion2:
                [npillars, nframes] = np.fromfile(f, dtype=dt_int, count=2)
                # nframes     = np.fromfile(f, dtype=dt_int, count=1)
                buffer = np.fromfile(f, dtype=dt_double, count=7)
                header.pixel_size = buffer[0]  # = np.fromfile(f, dtype=dt_double, count=1)
                header.diameter = buffer[1]  # = np.fromfile(f, dtype=dt_double, count=1)
                header.spacing = buffer[2]  # = np.fromfile(f, dtype=dt_double, count=1)
                header.oblique = buffer[3]  # = np.fromfile(f, dtype=dt_double, count=1)
                header.grid_angle = buffer[4]  # = np.fromfile(f, dtype=dt_double, count=1)
                header.sigma_PSF = buffer[5]  # = np.fromfile(f, dtype=dt_double, count=1)
                header.catch_radius = buffer[6]  # = np.fromfile(f, dtype=dt_double, count=1)
                [kernel_w, box_limit] = np.fromfile(f, dtype=dt_int, count=2)
                # kernel_w    = np.fromfile(f, dtype=dt_int, count=1)
                # box_limit   = np.fromfile(f, dtype=dt_int, count=1)
                header.kernel_w = kernel_w
                header.box_limit = box_limit

                buffer = np.fromfile(f, dtype=dt_int8, count=4)
                header.dark_pillar = buffer[0]  # np.fromfile(f, dtype=dt_int8, count=1)
                header.use_minimum_std = buffer[1]  # np.fromfile(f, dtype=dt_int8, count=1)
                header.use_metric_CG = buffer[2]  # np.fromfile(f, dtype=dt_int8, count=1)
                use_enhancer = buffer[3]  # np.fromfile(f, dtype=dt_int8, count=1)
                if version == fileversion2:
                    use_fft = np.fromfile(f, dtype=dt_int8, count=1)
                else:
                    use_fft = 0
                header.use_fft = use_fft
                header.use_enhancer = use_enhancer

                enhance_loaded_suc = True
                if use_enhancer > 0:
                    buffer = np.fromfile(f, dtype=dt_int8, count=3)
                    kernel.normalized_cc = buffer[0]  # np.fromfile(f, dtype=dt_int8, count=1)
                    kernel.use_gauss_psf = buffer[1]  # np.fromfile(f, dtype=dt_int8, count=1)
                    kernel.gauss_psf_dark = buffer[2]  # np.fromfile(f, dtype=dt_int8, count=1)
                    kernel.gauss_psf_radius = np.fromfile(f, dtype=dt_int, count=1)
                    kernel.gauss_psf_sigma = np.fromfile(f, dtype=dt_double, count=1)
                    [psf_w, psf_h] = np.fromfile(f, dtype=dt_int, count=1)
                    kernel.psf_w = psf_w
                    kernel.psf_h = psf_h
                    # psf_w = np.fromfile(f, dtype=dt_int, count=1)
                    # psf_h = np.fromfile(f, dtype=dt_int, count=1)
                    if psf_w > 0 and psf_h > 0:
                        PSF = np.fromfile(f, dtype=dt_double, count=psf_w * psf_h)
                        kernel.PSF = PSF.reshape(psf_h, psf_w)
                    else:
                        enhance_loaded_suc = False
                    header.kernel = kernel
                header.enhance_loaded_suc = enhance_loaded_suc

                if use_fft > 0:
                    fft_para = np.fromfile(f, dtype=dt_int, count=5)
                    fft.off_center_radius = fft_para[0]
                    fft.center_radius = fft_para[1]
                    fft.start_off_center_radius = fft_para[2]
                    fft.end_off_center_center_radius = fft_para[3]
                    num_points = fft_para[4]
                    if num_points > 0:
                        points = np.fromfile(f, dtype=dt_int, count=2 * num_points)
                        fft.points = points.reshape(2, num_points)
                    fft.num_points = num_points
                    header.fft = fft

                if enhance_loaded_suc:
                    startXY = np.fromfile(f, dtype=dt_int, count=2 * npillars)
                    header.startXY = startXY.reshape(2, npillars)
                    suc_load = True
            elif version % 2 == 0:
                npillars = version / 2
                nframes = np.fromfile(f, dtype=dt_int, count=1)
                suc_load = True
            # check whether the file header is loaded successfully.
            if suc_load and npillars > 0 and nframes > 0:
                ncols = npillars * 2
                nnsize = ncols * nframes
                A = np.fromfile(f, dtype=dt_double, count=nnsize)
                B = np.fromfile(f, dtype=dt_double, count=nnsize)
                A = A.reshape(ncols, nframes)
                B = B.reshape(ncols, nframes)
                end = ncols + 1;
                if version == fileversion2:
                    data.DX = B[0:end:2, :].transpose()
                    data.DY = B[1:end:2, :].transpose()
                    B = np.fromfile(f, dtype=dt_double, count=nnsize)
                    B = B.reshape(ncols, nframes)
                data.trackX = A[0:end:2, :].transpose()
                data.trackY = A[1:end:2, :].transpose()

                if B.size > 0:
                    data.drift_correctX = B[0:end:2, :].transpose()
                    data.drift_correctY = B[1:end:2, :].transpose()
                    driftXY = np.fromfile(f, dtype=dt_double, count=2 * nframes)
                    data.driftXY = driftXY.reshape(2, nframes)
                    data.reference_flags = np.fromfile(f, dtype=dt_int8, count=npillars)
            else:
                suc_load = False
                print('loading file failed.')

        header.suc_load = suc_load
        header.npillars = npillars
        header.nframes = nframes
        header.version = version
        return [header, data]
